numpyro/contrib/control_flow/scan.py

```diff
@@ -152,24 +149,33 @@ def body_fn(wrapped_carry, x, prefix=None):
                 # at each site have the same batch dims (e.g. if `fn.batch_shape = (2, 3)`,
                 # and value's batch_shape is (3,), then we promote shape of
                 # value so that its batch shape is (1, 3)).
-                new_carry, y = config_enumerate(seeded_fn)(carry, x)
+                with handlers.scope(divider='_'):
+                    new_carry, y = config_enumerate(seeded_fn)(carry, x)
 
             # store shape of new_carry at a global variable
-            nonlocal carry_shape_at_t1
-            carry_shape_at_t1 = [jnp.shape(x) for x in tree_flatten(new_carry)[0]]
+            if len(carry_shapes) < (history + 1):
+                carry_shapes.append([jnp.shape(x) for x in tree_flatten(new_carry)[0]])
             # make new_carry have the same shape as carry
             # FIXME: is this rigorous?
             new_carry = tree_multimap(lambda a, b: jnp.reshape(a, jnp.shape(b)),
                                       new_carry, carry)
-        return (i + jnp.array(1), rng_key, new_carry), (PytreeTrace(trace), y)
+        return (i + 1, rng_key, new_carry), (PytreeTrace(trace), y)
 
-    with markov():
+    with markov(history=history):
         wrapped_carry = (0, rng_key, init)
-        wrapped_carry, (_, y0) = body_fn(wrapped_carry, x0)
-        if length == 1:
-            ys = tree_map(lambda x: jnp.expand_dims(x, 0), y0)
-            return wrapped_carry, (PytreeTrace({}), ys)
-        wrapped_carry, (pytree_trace, ys) = lax.scan(body_fn, wrapped_carry, xs_, length - 1, reverse)
+        y0s = []
+        for i in range(history):
```

Comment:
For clarification, after changing this to `for i in range(2 * history)` what other changes would be necessary to get the correct final `carry_shape`?